<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [3]:
import pandas as pd
import psycopg2

In [5]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [15]:
# текст запроса
query_3_1 = f'''select count(*)
                from public.vacancies
             '''

In [16]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_7228\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [17]:
# текст запроса
query_3_2 = f'''select count(*)
                from public.employers
             '''

In [18]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_7228\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [19]:
# текст запроса
query_3_3 = f'''select count(*)
                from public.areas
             '''

In [20]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_7228\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [21]:
# текст запроса
query_3_4 = f'''select count(*)
                from public.industries
             '''

In [22]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_7228\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

## Выводы по предварительному анализу данных
- Количество вакансий в 2 раза больше, чем количество работодателей. Можно предположить, что у каждого работодателя есть как минимум 2 вакансии.
- Можно предположить, что на каждый регион приходится в среднем 30 вакансий.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [168]:
# текст запроса
query_4_1 = f'''select 
                    a.name,
                    count(v.id)
                from public.areas a
                join public.vacancies v on v.area_id=a.id
                group by a.name
                order by 2 desc
             '''

In [169]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_6948\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


name  count
0             Москва   5333
1    Санкт-Петербург   2851
2              Минск   2112
3        Новосибирск   2006
4             Алматы   1892
..               ...    ...
764       Тарко-Сале      1
765    Новоаннинский      1
766            Бирск      1
767           Сасово      1
768             Анна      1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [26]:
# текст запроса
query_4_2 = f'''select count(*)
                from public.vacancies
                where salary_from is not null or salary_to is not null
             '''

In [27]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_7228\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [28]:
# текст запроса
query_4_3 = f'''select 
                    round(avg(salary_from),0) as avg_salary_from,
                    round(avg(salary_to),0) as avg_salary_to
                from public.vacancies
             '''

In [29]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_7228\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [30]:
# текст запроса
query_4_4 = f'''select 
                    schedule,
                    employment,
                    count(id)
                from public.vacancies
                group by schedule, employment
                order by 3 desc
             '''

In [31]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_7228\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [32]:
# текст запроса
query_4_5 = f'''select 
                    experience,
                    count(id)
                from public.vacancies
                group by experience
                order by 2
             '''

In [33]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_7228\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

## Выводы по детальному анализу вакансий
- Средняя зарплата в вакансиях от 71тыс до 110 тыс.
- Больше половины от всех вакансий расчитаны на полный день и полную занятость. 
- В основном работадатели ищут людей с опытом работы от 1 до 3 лет (примерно половина вакансий от общего числа). Мало вакансий для людей с опытом более 6 лет. Это может быть связано с тем, что соискатели с большим опытом просят высокую заработную плату. 

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [199]:
# текст запроса
query_5_1 = f'''select 
                    e.name,
                    count(v.id)
                from public.employers e
                join public.vacancies v on v.employer_id = e.id
                group by e.name
                order by 2 desc
             '''

In [200]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_7228\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0                      Яндекс   1933
1                  Ростелеком    491
2                    Тинькофф    444
3                        СБЕР    428
4               Газпром нефть    331
...                       ...    ...
14761            им. Тьюринга      1
14762        LLC Novalab Tech      1
14763   Компания Мир упаковки      1
14764  Лифинцев Олег Иванович      1
14765      ТрансЭнергоХолдинг      1

[14766 rows x 2 columns]

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [37]:
# текст запроса
query_5_2 = f'''select 
                    a.name,
                    count(e.id) as count_employers,
                    count(v.id) as count_vacancies
                from public.areas a
                left join public.employers e on e.area=a.id
                left join public.vacancies v on v.area_id=a.id
                group by a.name
                having count(v.id)=0
                order by 2 desc
             '''

In [38]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df



C:\Users\Оля\AppData\Local\Temp\ipykernel_7228\898748541.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  count_employers  count_vacancies
0                          Россия              410                0
1                       Казахстан              207                0
2              Московская область               75                0
3              Краснодарский край               19                0
4              Ростовская область               18                0
..                            ...              ...              ...
588                      Струнино                0                0
589    Славгород (Алтайский край)                0                0
590                     Сорочинск                0                0
591  Ленинский (Тульская область)                0                0
592                         Тавда                0                0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [39]:
# текст запроса join public.areas a on v.area_id=a.id  
query_5_3 = f'''select 
                    e.name,
                    count(DISTINCT v.area_id) as count_area
                from public.vacancies v
                join public.employers e on e.id=v.employer_id
                group by e.name
                order by 2 desc
             '''



In [40]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df



C:\Users\Оля\AppData\Local\Temp\ipykernel_7228\2999002050.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  count_area
0                      Яндекс         181
1                  Ростелеком         152
2                  Спецремонт         116
3      Поляков Денис Иванович          88
4                   ООО ЕФИН           71
...                       ...         ...
14761                  UniSol           1
14762            UNISTORY LLC           1
14763                   UNIT6           1
14764     United Distribution           1
14765              UNITEL LLC           1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [41]:
# текст запроса
query_5_4 = f'''select 
                    count(e.id)
                from public.employers e
                left join public.employers_industries ei on e.id=ei.employer_id
                where ei.industry_id is null

             '''


In [42]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_7228\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [ ]:
# текст запроса
query_5_5 = f'''select 
                    e.name
                from public.employers e
                left join public.employers_industries ei on e.id=ei.employer_id
                group by e.id
                having count(ei.industry_id) = 4
                order by 1
                offset 2 limit 1        
             '''


In [208]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_7228\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [45]:
# текст запроса
query_5_6 = f'''select 
                    count(e.id)
                from public.employers e
                left join public.employers_industries ei on e.id=ei.employer_id
                left join public.industries i on i.id=ei.industry_id
                where i.name='Разработка программного обеспечения'    
             '''

In [46]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_7228\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [ ]:
# # код для получения списка городов-милионников
import pandas as pd
import requests # Импортируем библиотеку requests
from bs4 import BeautifulSoup # Импортируем библиотеку BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России' # Определяем адрес страницы
response = requests.get(url) # Делаем GET-запрос к ресурсу
print(response)
page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup
print(page.find('a')) # Ищем ссылку по тегу <a> и выводим её на экран

links = page.find_all('a') # Ищем все ссылки на странице и сохраняем в переменной links в виде списка
print(len(links)) # Выводим количество найденных ссылок

list_city = tuple([link.text for link in links[37:53]]) # Выводим список городов
print(list_city)

In [196]:
# текст запроса
query_5_7 = f'''select 
                     a.name,
                     count(v.id) cnt
                from public.vacancies v
                join public.areas a on a.id=v.area_id
                join public.employers e on e.id=v.employer_id
                where a.name in {list_city} and e.name='Яндекс'
                group by e.id, a.id
               
          
                Union all
                select 
                     'total',
                     count(v.id)
                from public.vacancies v
                join public.areas a on a.id=v.area_id
                join public.employers e on e.id=v.employer_id
                where a.name in {list_city} and e.name='Яндекс'
                order by 2 
                
                
                
             '''

In [197]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_7228\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7            Самара   26
8               Уфа   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            total  485

***

## Выводы по анализу работодателей
- Яндекс имеет самое большое количество опубликованных вакансий (всего 1933). Только среди городов-миллиоников у него вакансий 485. Также у Яндекса больше всех регионов, где имеются их вакансии.
- У 3553 работодателей сферой деятельности является Разработка программного обеспечения (это приблизительно 15%). 
- Всего работодателей 23501
- У 35% работадателей не указаны сферы деятельности. 



# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [209]:
# текст запроса
query_6_1 = f'''select 
                     count(v.id) cnt
                from public.vacancies v
                where lower(v.name) like '%data%' or lower(v.name) like '%данн%'
                
                
             '''

In [210]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_7228\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


cnt
0  1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [172]:
# текст запроса
query_6_2 = f'''select 
                    count(*)
                from public.vacancies v
                where (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (lower(v.name) like '%ml%' and lower(v.name) not like '%html%') or 
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                    ) and 
                    (lower(v.name) like '%junior%' or experience = 'Нет опыта' or employment = 'Стажировка')
                
                
             '''

In [173]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_6948\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [160]:
# текст запроса
query_6_3 = f'''select 
                    count(*)
                from public.vacancies v
                where (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (lower(v.name) like '%ml%' and lower(v.name) not like '%html%') or 
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                    ) 
                and (
                    lower(key_skills) like '%sql%' or 
                    lower(key_skills) like '%postgre%'
                    )
                
                
             '''


In [161]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_6948\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [154]:
# текст запроса
query_6_4 = f'''select 
                    count(*)
                from public.vacancies v
                where (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (lower(v.name) like '%ml%' and lower(v.name) not like '%html%') or 
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                    )  
                    and lower(key_skills) like '%python%'
                
                
             '''

In [155]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_6948\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [149]:
# текст запроса
query_6_5 = f'''select 
                    round(avg(length(v.key_skills)-length(replace(v.key_skills, CHR(9), ''))+1),2) as "среднее количество навыков"
                from public.vacancies v
                where
                    (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (v.name like '%ML%' and lower(v.name) not like '%html%') or 
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                    )  
                    AND v.key_skills IS NOT NULL
                    
                    
                    
                    
                
                
             '''

In [ ]:
# avg((length(v.key_skills)-(length(replace(v.key_skills, '\t', '')))+1)) as "среднее количество навыков"

In [150]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_6948\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


среднее количество навыков
0                        6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
# текст запроса
query_6_6 = f'''select 
                    v.experience as "опыт",
                    round(avg(coalesce((v.salary_from + v.salary_to)/2, v.salary_from, v.salary_to))) as "средняя зарплата"
                from public.vacancies v
                where (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (v.name like '%ML%' and lower(v.name) not like '%html%') or
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                    )  
                    and
                    (v.salary_from is not null or v.salary_to is not null)
                    group by v.experience
                      
                   
                    
                    
                    
                
                
             '''

In [157]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_6948\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


опыт  средняя зарплата
0           Нет опыта           74643.0
1  От 1 года до 3 лет          139675.0
2       От 3 до 6 лет          243115.0

***

## Выводы по предметному анализу
- Для вакансий DS без опыта минимальная средняя заработная плата от 74тыс. Причем вакансии рассчитанные на соискателей без опыта всего 10% от всего количества вакансий по DS. 
- Для вакансий DS требуется в среднем минимум 6 ключевых навыков. Основные из них Python, SQL. SQL требуется в почти 50% вакансиях, python в почти 70% вакансиях.

# Общий вывод по проекту

## Дополнительные вычисления

In [1]:
# Определить зависимость количества вакансий DS от опыта работы
query_7_1 = f'''select 
                    v.experience as "опыт",
                    count(v.id)
                from public.vacancies v
                where (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (v.name like '%ML%' and lower(v.name) not like '%html%') or
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                    )
                    group by v.experience
                      
                   
                    
                    
                    
                
                
             '''

In [6]:
df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_6164\2911634181.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


опыт  count
0         Более 6 лет     34
1           Нет опыта     28
2  От 1 года до 3 лет    223
3       От 3 до 6 лет    195

In [8]:
# Сколько вакансий приходится на работадателей без сферы деятельности. 
query_7_2 = f'''select 
                    count(v.id)
                from public.employers e
                left join public.employers_industries ei on e.id=ei.employer_id
                join public.vacancies v on v.employer_id=e.id
                where ei.industry_id is null

             '''


In [9]:
df = pd.read_sql_query(query_7_2, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_3460\2993136978.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2, connection)


count
0  10084

In [14]:
# Определить среднюю заработную плату в вакансиях DS в зависимости от региона.
query_7_3 = f'''select 
                    a.name as "регион",
                    round(avg(coalesce((v.salary_from + v.salary_to)/2, v.salary_from, v.salary_to))) as "средняя зарплата"
                from public.vacancies v
                join public.areas a on a.id=v.area_id
                where (
                    lower(v.name) like '%data scientist%' or 
                    lower(v.name) like '%data science%' or 
                    lower(v.name) like '%исследователь данных%' or 
                    (v.name like '%ML%' and lower(v.name) not like '%html%') or
                    lower(v.name) like '%machine learning%' or 
                    lower(v.name) like '%машинн%обучен%'
                    )  
                    and
                    (v.salary_from is not null or v.salary_to is not null)
                    group by a.name
                    order by 2 desc
                
             '''

In [15]:
df = pd.read_sql_query(query_7_3, connection)
df

C:\Users\Оля\AppData\Local\Temp\ipykernel_6164\2328698441.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_3, connection)


регион  средняя зарплата
0              Кипр          300000.0
1           Армения          268863.0
2        Черногория          233794.0
3            Турция          233794.0
4            Сербия          233794.0
5            Москва          207630.0
6          Белгород          200000.0
7   Санкт-Петербург          173335.0
8       Новосибирск          166333.0
9            Рязань          136667.0
10          Королев          130000.0
11     Екатеринбург          124000.0
12           Алматы           94587.0
13              Уфа           90000.0
14      Владивосток           85000.0
15            Пермь           75000.0
16          Барнаул           60000.0
17   Ростов-на-Дону           50000.0
18           Казань           50000.0

In [16]:
# Закрываем соединение
connection.close()

## Общие выводы

1. Из всех вакансий по DS 78% вакансий, причем почти в половине всех вакансий требуется опыт от одного года до трёх лет. Если поделить на условные грейды, то это специалисты Junior+ и Middle. Учитывая текущий повышенный интерес к направлению нейросетей во многих компаниях, не удивительно, что требуются специалисты именно с таким стажем – они основная рабочая сила, часто идеальный баланс между запрашиваемой зарплатой и скиллами.
Специалистов же с опытом более 6 лет требуется мало, так как не многим работодателям нужен такой уровень экспертности у специалиста, и это могут быть, например менеджеры среднего и высшего звена, руководители направлений. Подобных вакансий всегда не очень много на рынке.

2. Сфера деятельности работодателя не заполнена в 2% общего числа вакансий. Если бы в базе данных была информация о сроке жизни вакансий, можно было бы проверить, как эти два показателя коррелируют между собой. Но так как такой информации нет, могу только сделать предложение, что эти вакансии могут дольше висеть незакрытыми, так как могут не попадать в поисковые выдачи части соискателей.


Большая часть анализа была направлена на вакансии по DS. Видно, что направление бурно растет,  вакансии есть не только в крупных городах. Интересно было бы теми же запросами проанализировать текущую базу вакансий ХХ, и сравнить, насколько общие по рынку тенденции на охлаждение найма повлияли конкретно на направление DS.